tests/exchange/test_gateio.py

```diff
@@ -26,3 +29,48 @@ def test_validate_order_types_gateio(default_conf, mocker):
     with pytest.raises(OperationalException,
                        match=r'Exchange .* does not support market orders.'):
         ExchangeResolver.load_exchange('gateio', default_conf, True)
+
+
+@pytest.mark.parametrize('pair,mm_ratio', [
+    ("ETH/USDT:USDT", 0.005),
+    ("ADA/USDT:USDT", 0.003),
+    ("DOGE/USDT:USDT", None),
+])
+def test_get_maintenance_ratio_and_amt_gateio(default_conf, mocker, pair, mm_ratio):
+    mocker.patch(
+        'freqtrade.exchange.Exchange.markets',
+        PropertyMock(
+            return_value={
+                'ETH/USDT:USDT': {
+                    'taker': 0.0000075,
+                    'maker': -0.0000025,
+                    'info': {
+                        'maintenance_rate': '0.005',
+                    },
+                    'id': 'ETH_USDT',
+                    'symbol': 'ETH/USDT:USDT',
+                },
+                'ADA/USDT:USDT': {
+                    'taker': 0.0000075,
+                    'maker': -0.0000025,
+                    'info': {
+                        'maintenance_rate': '0.003',
+                    },
+                    'id': 'ADA_USDT',
+                    'symbol': 'ADA/USDT:USDT',
+                },
+                'DOGE/USDT:USDT': {
+                    'taker': 0.0000075,
+                    'maker': -0.0000025,
+                    'info': {
+                        'nonmaintenance_rate': '0.003',
+                    },
+                    'id': 'DOGE_USDT',
+                    'symbol': 'DOGE/USDT:USDT',
+                }
+            }
+        )
+    )
+    api_mock = MagicMock()
+    exchange = get_patched_exchange(mocker, default_conf, api_mock, id="gateio")
+    assert exchange.get_maintenance_ratio_and_amt(pair) == [mm_ratio, None]
```

Comment:
@xmatthias So this test keeps failing because exchange.markets doesn't get set to what I have written there. It gets set the value created in `conftest.py`

```
{'BLK/BTC': {'active': True,
             'base': 'BLK',
             'id': 'blkbtc',
             'limits': {'amount': {'max': 1000, 'min': 0.01},
                        'cost': {'max': 500000, 'min': 0.0001},
                        'leverage': {'max': 3.0, 'min': 1.0},
                        'price': 500000},
             'lot': 1e-08,
             'precision': {'amount': 8, 'cost': 8, 'price': 8},
             'quote': 'BTC',
             'spot': True,
             'symbol': 'BLK/BTC',
             'type': 'spot'},
 'BTT/BTC': {'active': False,
             'base': 'BTT',
             'id': 'BTTBTC',
             'info': {},
             'limits': {'amount': {'max': 90000000.0, 'min': 1.0},
                        'cost': {'max': None, 'min': 0.0001},
                        'price': {'max': None, 'min': None}},
             'precision': {'amount': 0, 'base': 8, 'price': 8, 'quote': 8},
             'quote': 'BTC',
             'spot': True,
             'symbol': 'BTT/BTC',
             'type': 'spot'},
 'ETH/BTC': {'active': True,
             'base': 'ETH',
             'id': 'ethbtc',
             'limits': {'amount': {'max': 1000, 'min': 0.01},
                        'cost': {'max': 500000, 'min': 0.0001},
                        'leverage': {'max': 2.0, 'min': 1.0},
                        'price': 500000},
             'lot': 1e-08,
             'precision': {'amount': 8, 'cost': 8, 'price': 8},
             'quote': 'BTC',
             'spot': True,
             'symbol': 'ETH/BTC',
             'type': 'spot'},
 'ETH/USDT': {'active': True,
              'base': 'ETH',
              'future': True,
              'id': 'USDT-ETH',
              'info': {},
              'limits': {'amount': {'max': None, 'min': 0.02214286},
                         'price': {'max': None, 'min': 1e-08}},
              'margin': True,
              'precision': {'amount': 8, 'price': 8},
              'quote': 'USDT',
              'spot': True,
              'swap': True,
              'symbol': 'ETH/USDT',
              'type': 'spot'},
 ```
 
 Do you know why it's not getting set correctly?

Reply:
`get_patched_exchange` patches markets itself (and will therefore overwrite your mock).

You should be able to simply reverse the order (do your mock after getting the exchange object) ...